src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DefaultLSPDocumentMappingProvider.cs

```diff
@@ -58,5 +70,156 @@ public async override Task<RazorMapToDocumentRangeResponse> MapToDocumentRangeAs
+                    // Couldn't find the document. Ignore this edit.
+                    continue;
+                }
+
+                remappedDocumentEdits.Add(new TextDocumentEdit()
+                {
+                    TextDocument = new VersionedTextDocumentIdentifier()
+                    {
+                        Uri = documentSnapshot.Uri,
+                        Version = documentSnapshot.Version
+                    },
+                    Edits = remappedEdits
+                });
+            }
+
+            return remappedDocumentEdits.ToArray();
+        }
+
+        private async Task<Dictionary<string, TextEdit[]>> RemapDocumentEditsAsync(Dictionary<string, TextEdit[]> changes, CancellationToken cancellationToken)
+        {
+            var remappedChanges = new Dictionary<string, TextEdit[]>();
+            foreach (var entry in changes)
+            {
+                var uri = new Uri(entry.Key);
+                var edits = entry.Value;
+
+                if (!CanRemap(uri))
+                {
+                    // This location doesn't point to a background razor file. No need to remap.
+                    remappedChanges[entry.Key] = entry.Value;
+                    continue;
+                }
+
+                var (documentSnapshot, remappedEdits) = await RemapTextEditsAsync(uri, edits, cancellationToken).ConfigureAwait(false);
+                if (documentSnapshot == null)
+                {
+                    // Couldn't find the document. Ignore this edit.
+                    continue;
+                }
+
+                remappedChanges[documentSnapshot.Uri.GetAbsoluteOrUNCPath()] = remappedEdits;
+            }
+
+            return remappedChanges;
+        }
+
+        private async Task<(LSPDocumentSnapshot, TextEdit[])> RemapTextEditsAsync(Uri uri, TextEdit[] edits, CancellationToken cancellationToken)
+        {
+            var languageKind = RazorLanguageKind.Razor;
+            if (RazorLSPConventions.IsRazorCSharpFile(uri))
+            {
+                languageKind = RazorLanguageKind.CSharp;
+            }
+            else if (RazorLSPConventions.IsRazorHtmlFile(uri))
+            {
+                languageKind = RazorLanguageKind.Html;
+            }
+            else
+            {
+                Debug.Fail("This method should only be called for Razor background files.");
+            }
+
+            var razorDocumentUri = RazorLSPConventions.GetRazorDocumentUri(uri);
+            if (!_documentManager.TryGetDocument(razorDocumentUri, out var documentSnapshot))
+            {
+                return (null, EmptyEdits);
+            }
+
+            var remappedEdits = new List<TextEdit>();
+            foreach (var edit in edits)
+            {
+                var mappingResult = await MapToDocumentRangeAsync(
+                languageKind,
+                razorDocumentUri,
+                edit.Range,
+                cancellationToken).ConfigureAwait(false);
+
```

Comment:
Lets throw some cancellation checks in here to ensure we're not doing any extra work. I'm also assuming that checking if cancellation is requested is cheap (@halter73 is that a fair assumption from what you've seen?)

Reply:
I did that in `RenameHandler` already. I wasn't sure if we want it here given that we plan to batch our maptodocumentrange requests in the future.

Reply:
ThrowIfCancellationRequested() is really cheap. If the cancellation token isn't cancellable, it's just an inline null check on a struct field. If it is cancellable, it has to check the backing CTS, but that's still quite fast. We do it for every read and write in Kestrel.

Reply:
> I did that in `RenameHandler` already. I wasn't sure if we want it here given that we plan to batch our maptodocumentrange requests in the future.

Fair enough